### ISSUE TITLE
Werkzeug stat reloader problems

### ISSUE THREAD

**Issue by untitaker** [mitsuhiko/werkzeug#682]
So far two people have run into serious performance issues with the stat reloader. Anybody who has those, **the fix is to install watchdog from pypi**.

If you're interested in fixing this issue, uninstall watchdog again and:

* How does setting ``reloader_interval`` on ``werkzeug.run_simple`` or ``app.run`` affect performance? Try various values:

        app.run(reloader_interval=1)  # the default

* Run this code inside your virtualenv, post the returned number:

        from werkzeug._reloader import _iter_module_files
        print(len(list(_iter_module_files())))

**Comment by untitaker** [comment]
cc @joostdevries

**Comment by txtsd** [comment]
Increasing the reloader_interval only increases the time between cpu usage spikes. With watchdog installed, the cpu usage is almost non-existent!
```
┌[txtsd@dungeon-of-data]─[~/sites/flasktut] 
└─▶ flask/bin/python
Python 3.4.2 (default, Jan 12 2015, 11:38:40) 
[GCC 4.9.2 20141224 (prerelease)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from werkzeug._reloader import _iter_module_files
>>> print(len(list(_iter_module_files())))
10893
>>> 
```

**Comment by untitaker** [comment]
I think it'd be sensible to re-introduce the old behavior of only scanning ``sys.modules`` for the ``stat`` reloader. It is incorrect, but more importantly, it is fast without any dependencies. If people encounter issues with unwatched files (as in #569), we can simply tell them to use watchdog.

**Comment by joostdevries** [comment]
@untitaker I'd just recommend people to use `reloader_paths`. In quite a lot of cases, `sys.modules` will still yield a huge list of files. I do think there's a regression in the most recent version, in my case a CPU usage increase of about 30%.

**Comment by untitaker** [comment]
@joostdevries ``reloader_paths`` is a workaround to a problem that shouldn't have been introduced in the first place. The difference in the algorithm for finding the files to watch is more dramatic than you might think: Werkzeug has to traverse all directories recursively, launching ``stat`` for anything beneath them.

**Comment by alanhamlett** [comment]
I noticed this issue right away when my laptop battery lasted only a few hours because werkzeug was constantly taking 20% CPU. Installing watchdog fixed the issue, but would be great to not have the default have such horrible performance...

**Comment by untitaker** [comment]
Ok, probably makes sense to reintroduce the old behavior. iter-module-files yields 200k items for me and I haven't noticed the perf issue on my desktop.

**Comment by RonnyPfannschmidt** [comment]
i do wonder if it would be so bad to depend on a optimized package for reloading in general in any case

**Comment by untitaker** [comment]
On Thu, Mar 05, 2015 at 07:47:06AM -0800, Ronny Pfannschmidt wrote:
> i do wonder if it would be so bad to depend on a optimized package for
> reloading in general in any case

Some filesystems don't have filesystem events (there are open issues about some
setup in Vagrant that is affected by this). I think Watchdog also has a stat
reloader available, but I don't think it's worth depending on Watchdog for that
one.

**Comment by RonnyPfannschmidt** [comment]
in that case i do wonder if the relaoding behavior could be enhanced if stuff like dot-files/hidden files and virtualenvs could be ignored along with anything not in the cwd

**Comment by rogerntan** [comment]
Hi, I ran into this thread after we noticed high CPU usage figures on idle (30-50%). I tried the watchdog suggestion which worked fine initially, but after a short period of time no pages would be served up by the web server, even after subsequent process restarts and killing zombie python processes.

**Comment by ya-mouse** [comment]
I have another problem with stat reloader. With version 0.10 if I run this simple app:
```python
from flask import Flask
import requests
app = Flask(__name__)
@app.route('/')
def index():
     return 'Index'
app.run(debug=True, host='127.0.0.1')
```
then I've got the following traceback:
```
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
Traceback (most recent call last):
  File "app.py", line 7, in <module>
    app.run(debug = True, host='127.0.0.1')
  File "/usr/local/lib/python3.4/dist-packages/flask/app.py", line 772, in run
    run_simple(host, port, self, **options)
  File "/usr/local/lib/python3.4/dist-packages/werkzeug/serving.py", line 622, in run_simple
    reloader_type)
  File "/usr/local/lib/python3.4/dist-packages/werkzeug/_reloader.py", line 269, in run_with_reloader
    reloader.run()
  File "/usr/local/lib/python3.4/dist-packages/werkzeug/_reloader.py", line 159, in run
    for filename in chain(_iter_module_files(), self.extra_files):
  File "/usr/local/lib/python3.4/dist-packages/werkzeug/_reloader.py", line 70, in _iter_module_files
    for package_path in getattr(module, '__path__', ()):
  File "/usr/lib/python3/dist-packages/six.py", line 116, in __getattr__
    _module = self._resolve()
  File "/usr/lib/python3/dist-packages/six.py", line 105, in _resolve
    return _import_module(self.mod)
  File "/usr/lib/python3/dist-packages/six.py", line 76, in _import_module
    __import__(name)
ImportError: No module named 'winreg'
```
If I downgrade to 0.9 or remove "import requests" everything would be fine.

**Comment by untitaker** [comment]
Everybody who has this issue:

With the stat-reloader, could you delete/comment out the following three lines in `werkzeug/_reloader`:

    for package_path in getattr(module, '__path__', ()):
        for filename in _recursive_walk(os.path.abspath(package_path)):
            yield filename

How much does that affect the overall performance of the `stat`-reloader?

**Comment by untitaker** [comment]
@ya-mouse Are you sure this is related to Flask or Werkzeug at all?

**Comment by rogerntan** [comment]
@untitaker On Werkzeug 0.10.1, after commenting out those 3 lines, I'm still seeing high CPU usage similar to the usage with the lines intact.

**Comment by ya-mouse** [comment]
@untitaker I think so. Newer Werkzeug version tries to get __path__ attribute from six.MovedModule instance (mapping for "tkinter.commondialog") or mapping for "winreg" if python3-tk has been installed.
Not all values in sys.modules are real modules. So, this problem related to Werkzeug 0.10.1.

**Comment by peplin** [comment]
@ya-mouse I had this problem too, it's an incompatibility between many auto-reloaders (Werkzeug's, Django's, etc) and older versions of `six`. I was able to fix it with this monkey patch in my project: https://gist.github.com/peplin/88ef5f26c4cc53160448

**Comment by untitaker** [comment]
Regarding the six-issue: Please open a separate issue for this.

**Comment by Lothiraldan** [comment]
The six issue has been already opened and was just fixed one hour ago: https://bitbucket.org/gutworth/six/issue/99/six-and-inspect-importerror-_winreg-module

**Comment by untitaker** [comment]
Everybody who reported issues with the reloader, could you try out https://github.com/mitsuhiko/werkzeug/pull/715 ?

**Comment by davidkhess** [comment]
I'm seeing the increased CPU usage also - this is through a VirtualBox VM of Ubuntu 14.04.

I would be happy to install watchdog but watchdog doesn't seem to support gevent. (Would like to be proven wrong on that point though.)

**Comment by untitaker** [comment]
Please try out #715

**Comment by davidkhess** [comment]
Do you happen to have the pip magic handy to make that an easy uninstall/install?

**Comment by untitaker** [comment]
pip install git+https://github.com/untitaker/werkzeug.git#reloader-perf

**Comment by davidkhess** [comment]
Thanks!

That branch appears to work. Reloader is reloading, requests are being served and the CPU usage is back down to something reasonable.

**Comment by ya-mouse** [comment]
@Lothiraldan the six-related issue just fix a half of problem. If I don't have *tkinter*, then Werkzeug would fail again and again. I've opened separate issue #717.

**Comment by untitaker** [comment]
@davidkhess 

>However, on another VM with Werkzeug 0.9.4, I'm seeing less than 1% CPU usage. Different application so the modules to scan are going to be different but still another data point.

Not sure if I can do anything with that data. It'd be great if you could install the patched Werkzeug on *that* machine and see how it works out.

**Comment by davidkhess** [comment]
Ok, I put that branch on the 0.9.4 VM and CPU usage rose to around 5%.

**Comment by untitaker** [comment]
That's quite irritating because the same exact algorithm is used. However, this also could be because pip is installing from a repo. I guess we'll have to deal with it in either case.

**Comment by untitaker** [comment]
@joostdevries I don't think your reloader_paths feature makes sense given that the stat reloader now operates on files, as was the case before.

To see what I mean, try merging `0.10-maintenance` into master and resolving the conflicts. I'm not sure how `iter_module_files` should process `reloader_paths`.

**Comment by davidkhess** [comment]
@untitaker, I did an strace of both. The difference is your patched version is reading directory entries while 0.9.4 does not.

Your branch:

```
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py", {st_mode=S_IFREG|0664, st_size=28693, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/sqlalchemy/sql/dml.pyc", {st_mode=S_IFREG|0664, st_size=30625, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/sqlalchemy/sql/expression.py", {st_mode=S_IFREG|0664, st_size=5668, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython/error.so", {st_mode=S_IFREG|0775, st_size=60455, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython/error.so", {st_mode=S_IFREG|0775, st_size=60455, ...}) = 0
12:09:44 openat(AT_FDCWD, "/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 21
12:09:44 getdents(21, /* 19 entries */, 32768) = 592
12:09:44 getdents(21, /* 0 entries */, 32768) = 0
12:09:44 close(21)                      = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython/__init__.py", {st_mode=S_IFREG|0664, st_size=1469, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython/__init__.py", {st_mode=S_IFREG|0664, st_size=1469, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython/__init__.pyc", {st_mode=S_IFREG|0664, st_size=685, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/asn1.pyc", {st_mode=S_IFREG|0664, st_size=14265, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/asn1.py", {st_mode=S_IFREG|0664, st_size=12727, ...}) = 0
12:09:44 openat(AT_FDCWD, "/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 21
12:09:44 getdents(21, /* 24 entries */, 32768) = 800
12:09:44 getdents(21, /* 0 entries */, 32768) = 0
12:09:44 close(21)                      = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/Counter.py", {st_mode=S_IFREG|0664, st_size=5157, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/Counter.py", {st_mode=S_IFREG|0664, st_size=5157, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/randpool.pyc", {st_mode=S_IFREG|0664, st_size=3440, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/randpool.py", {st_mode=S_IFREG|0664, st_size=2768, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/asn1.py", {st_mode=S_IFREG|0664, st_size=12727, ...}) = 0
12:09:44 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/py21compat.py", {st_mode=S_IFREG|0664, st_size=2779, ...}) = 0
```

vs. 0.9.4 (not exactly the same but covering some common files)

```
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/werkzeug/contrib/__init__.pyc", {st_mode=S_IFREG|0664, st_size=781, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/werkzeug/contrib/__init__.py", {st_mode=S_IFREG|0664, st_size=623, ...}) = 0
12:10:32 stat("/home/databakery/app/app/models/users.pyc", {st_mode=S_IFREG|0664, st_size=14844, ...}) = 0
12:10:32 stat("/home/databakery/app/app/models/users.py", {st_mode=S_IFREG|0664, st_size=11388, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/jinja2/optimizer.pyc", {st_mode=S_IFREG|0664, st_size=2959, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/jinja2/optimizer.py", {st_mode=S_IFREG|0664, st_size=2302, ...}) = 0
12:10:32 stat("/usr/lib/python2.7/email/mime/__init__.pyc", {st_mode=S_IFREG|0644, st_size=128, ...}) = 0
12:10:32 stat("/usr/lib/python2.7/email/mime/__init__.py", {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython/constants.so", {st_mode=S_IFREG|0775, st_size=153384, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/zmq/backend/cython/constants.so", {st_mode=S_IFREG|0775, st_size=153384, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.pyc", {st_mode=S_IFREG|0664, st_size=46623, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", {st_mode=S_IFREG|0664, st_size=37539, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/PIL/__init__.pyc", {st_mode=S_IFREG|0664, st_size=1262, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/PIL/__init__.py", {st_mode=S_IFREG|0664, st_size=1523, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/sqlalchemy/connectors/zxJDBC.pyc", {st_mode=S_IFREG|0664, st_size=2741, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/sqlalchemy/connectors/zxJDBC.py", {st_mode=S_IFREG|0664, st_size=1868, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/Counter.pyc", {st_mode=S_IFREG|0664, st_size=4258, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/Crypto/Util/Counter.py", {st_mode=S_IFREG|0664, st_size=5157, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/rauth/utils.pyc", {st_mode=S_IFREG|0664, st_size=3206, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/rauth/utils.py", {st_mode=S_IFREG|0664, st_size=1959, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/jinja2/nodes.pyc", {st_mode=S_IFREG|0664, st_size=44770, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/jinja2/nodes.py", {st_mode=S_IFREG|0664, st_size=28875, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/requests/packages/urllib3/response.pyc", {st_mode=S_IFREG|0664, st_size=10846, ...}) = 0
12:10:32 stat("/home/databakery/flask/local/lib/python2.7/site-packages/requests/packages/urllib3/response.py", {st_mode=S_IFREG|0664, st_size=10537, ...}) = 0
12:10:32 stat("/usr/lib/python2.7/quopri.pyc", {st_mode=S_IFREG|0644, st_size=6550, ...}) = 0
```

**Comment by untitaker** [comment]
That's weird, I can't see anything where the code would be doing any directory access.

**Comment by untitaker** [comment]
To clarify, the reloader from 0.10 isn't by chance using watchdog?

**Comment by davidkhess** [comment]
If that question was directed at me, watchdog has not been installed during these tests.

**Comment by davidkhess** [comment]
Figured it out. Typo in pip install command:

`pip install git+https://github.com/untitaker/werkzeug.git@reloader-perf`

I.e. `@` instead of `#`

I guess it was loading master of your repo. Spec'd that way, it is back down to around 1% CPU.

**Comment by untitaker** [comment]
That's weird too, because installing from master should have the same behavior as 0.10.

Anyway, I think this is fixed, except for the broken reloader_paths feature. @joostdevries, I'm tempted to remove this feature from master again, since the cause of your problems has been fixed. Are you satisfied with that solution?

**Comment by untitaker** [comment]
This issue is fixed with 0.10.2.

@joostdevries Please refile your PR if you still need this functionality with 0.10.2, but I don't think you will need it.
